lib/active_merchant/billing/gateways/safe_charge.rb

```diff
@@ -146,12 +146,19 @@ def add_transaction_data(trans_type, post, money, options)
       end
 
       def add_payment(post, payment, options = {})
-        post[:sg_NameOnCard] = payment.name
-        post[:sg_CardNumber] = payment.number
         post[:sg_ExpMonth] = format(payment.month, :two_digits)
         post[:sg_ExpYear] = format(payment.year, :two_digits)
         post[:sg_CVV2] = payment.verification_value
-        post[:sg_StoredCredentialMode] = (options[:stored_credential_mode] == true ? 1 : 0)
+        post[:sg_CardNumber] = payment.number
+
+        if payment.is_a?(NetworkTokenizationCreditCard) && payment.source == :network_token
+          post[:sg_CAVV] = payment.payment_cryptogram
+          post[:sg_ECI] = options[:three_d_secure][:eci] || '05'
```

Comment:
@DStoyanoff - I think the logic for the following line of code is wrong:
`post[:sg_ECI] = options[:three_d_secure][:eci] || '05'`

If the value for options[:three_d_secure][:eci] is not set, it returns a Nil error instead of defaulting to a value of '05'

Please correct it asap. Since I have already merged this PR, you will have to create a new branch and then create a new PR for the fix.

Reply:
Hello @jimilpatel24. I have opened a new PR - https://github.com/activemerchant/active_merchant/pull/3920
Strangely it shows more diff then expected, as if this PR was not merged.. Let me know if there is anything else I can do